sale_coupon_multi_gift/models/sale_order.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2021 Tecnativa - David Vidal
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, models
+from odoo.tests import Form
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _get_paid_order_lines(self):
+        """Add reward lines produced by multi gift promotions"""
+        lines = super()._get_paid_order_lines()
+        free_reward_product = (
+            self.env["sale.coupon.program"]
+            .search([("reward_type", "=", "multi_gift")])
+            .mapped("coupon_multi_gift_ids.reward_product_id")
+        )
+        free_reward_product_lines = self.order_line.filtered(
+            lambda x: x.is_reward_line and x.product_id in free_reward_product
+        )
+        return lines | free_reward_product_lines
+
+    def _get_reward_values_multi_gift_line(self, reward_line, program):
+        """Multi Gift reward rules. For every gift reward rule, we'll create a new
+        sale order line flagged as reward line with a 100% discount"""
+        # Compute the proper price_unit to create the new line accordingly.
+        # We use the form class to ensure the proper onchanges.
+        sale_form = Form(self)
```

Comment:
This is going to be very slow when the order contains a lot of lines. I'm thinking on 2 techniques:

1. Create a form from scratch with the same partner
2. Do a `new` sale.order.line and call `onchange` method for the product/quantity fields.

How is this done on core?

Reply:
In core an automatic service product created for each promotion. As the product to enjoy the promotion must be added by the user prior to enjoy the discount we already have the line

https://github.com/odoo/odoo/blob/13.0/addons/sale_coupon/models/sale_order.py#L81

Then they take the easy way:

https://github.com/odoo/odoo/blob/13.0/addons/sale_coupon/models/sale_order.py#L107

Reply:
Change. I took the same approach as in `sale_elaboration` with a new() and a helper method to trigger the onchanges